### PR TITLE
oauth2 filter: handle large size cookies

### LINKED ISSUE #27875: oauth2 filter: browser blocks BearerToken cookie because its size too large
*Title*: oauth2 filter: browser blocks BearerToken cookie because its size too large

*Description*:

I expected the cookie for `BearerToken` to be set on the response when `forward_bearer_token: true`. Instead, the browser (Chrome Version 114.0.5735.106) returns an error message on the attempt:

```
The attempt to set a cookie via a Set-Cookie header was blocked because the cookie was too large. The combined size of the name and value must be less than or equal to 4096 characters. 
```

Then, the behavior is similar to https://github.com/envoyproxy/envoy/issues/21819, where the browser enters a 302 redirect loop.

It would be an improvement if the oauth2 filter could do anyone of the following:

1. Log a warning if the cookie size is too large
2. Allow one to only set the Authorization header and not use cookies
3. Set multiple cookies, each less than the max size, for a large BearerToken, e.g. splitting it up into parts that could be reassembled by the consumer

*Repro steps*:

With `auth_scopes` set to only `openid`, the *id token* is a reasonable size. With cloud providers such as Azure, when adding additional scopes for Microsoft applications, e.g. `https://cognitiveservices.azure.com/.default`, the *access token* returned seems to return a list of `groups`, which for some users can be quite large. Microsoft doesn't seem to allow customization of this token.

*Config*:

Using `envoy:v1.26.1`

```
                      - name: envoy.filters.http.oauth2
                        typed_config:
                          "@type": type.googleapis.com/envoy.extensions.filters.http.oauth2.v3.OAuth2
                          config:
                            token_endpoint:
                              cluster: <omitted>
                              uri: https://login.microsoftonline.com/<omitted>/oauth2/v2.0/token
                              timeout: 10s
                            authorization_endpoint: https://login.microsoftonline.com/<omitted>/oauth2/v2.0/authorize
                            redirect_uri: "https://%REQ(:authority)%/callback"
                            redirect_path_matcher:
                              path:
                                prefix: /callback
                            pass_through_matcher:
                              name: authorization
                              present_match: true
                            signout_path:
                              path:
                                exact: /signout
                            forward_bearer_token: true
                            credentials:
                              client_id: "<omitted>"
                              token_secret:
                                name: token
                                sds_config:
                                  resource_api_version: V3
                                  path_config_source:
                                    path: "/etc/envoy/token-secret.yaml"
                              hmac_secret:
                                name: hmac
                                sds_config:
                                  resource_api_version: V3
                                  path_config_source:
                                    path: "/etc/envoy/hmac-secret.yaml"
                            auth_scopes:
                              - openid
                              - <any scope for a Microsoft application>

```

*Logs*:

The 302 redirect loop will eventually end with:

```
[2023-06-08 13:09:06.628][51][debug][oauth2] [source/extensions/filters/http/oauth2/oauth_client.cc:88] Oauth response body: {"error":"invalid_grant","error_description":"AADSTS54005: OAuth2 Authorization code was already redeemed, please retry with a new valid code or use an existing refresh token.\r\nTrace ID: <omitted>r\nCorrelation ID: <omitted>\r\nTimestamp: 2023-06-08 13:09:06Z","error_codes":[54005],"timestamp":"2023-06-08 13:09:06Z","trace_id":"<omitted>","correlation_id":"<omitted>"}
```

### PR DESCRIPTION
**Pull request by kanurag94**
Commit Message: oauth2 filter: handle setting large sized cookies for response tokens
Additional Description: if `forward_bearer_token` is set to true, the user expects to use the token in further flows. If cookie size is greater than 4096, most modern browsers will reject it. Thus in this PR, we split the cookie in chunks of size 4000, and later recombine them during read
Risk Level: Low
Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
Fixes #27875
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

File: source/extensions/filters/http/oauth2/filter.cc [hunk 1]
```diff
@@ -517,6 +543,36 @@ void OAuth2Filter::finishGetAccessTokenFlow() {
   config_->stats().oauth_success_.inc();
 }
 
+void OAuth2Filter::setCookie(Http::ResponseHeaderMap& headers, const std::string& key,
+                             const std::string& value) const {
+  headers.addReferenceKey(Http::Headers::get().SetCookie, absl::StrCat(key, "=", value));
+}
+
+void OAuth2Filter::setSplitCookie(Http::ResponseHeaderMap& headers, const std::string& key,
+                                  const std::string& data, size_t maxChunkSize = 4000) const {
```
Comment (vikaschoudhary16) [inline source/extensions/filters/http/oauth2/filter.cc:552]: setting default to 4000 should be in the function declaration I think. Rather,  should we make it configuration option?

File: source/extensions/filters/http/oauth2/filter.cc [hunk 2]
```diff
@@ -517,6 +543,36 @@ void OAuth2Filter::finishGetAccessTokenFlow() {
...
+  if (data.size() <= maxChunkSize) {
+    // If the data is smaller than the maximum chunk size, no need to split
+    setCookie(headers, key, data);
+    return;
+  }
+
+  if (data.size() > maxChunkSize) {
+    ENVOY_LOG(debug, "cookie size is greater than {}: \n{}", maxChunkSize, key);
```
Comment (vikaschoudhary16) [inline source/extensions/filters/http/oauth2/filter.cc:560]: nit: I would prefer to print key in the same line. something like `cookie, {}, size is....`

File: source/extensions/filters/http/oauth2/filter.cc [hunk 3]
```diff
@@ -517,6 +543,36 @@ void OAuth2Filter::finishGetAccessTokenFlow() {
   config_->stats().oauth_success_.inc();
 }
 
+void OAuth2Filter::setCookie(Http::ResponseHeaderMap& headers, const std::string& key,
+                             const std::string& value) const {
+  headers.addReferenceKey(Http::Headers::get().SetCookie, absl::StrCat(key, "=", value));
+}
+
+void OAuth2Filter::setSplitCookie(Http::ResponseHeaderMap& headers, const std::string& key,
```
Comment (vikaschoudhary16) [inline source/extensions/filters/http/oauth2/filter.cc:551]: nit: setSplitCookie/setCookieOrChunkedCookies?

File: source/extensions/filters/http/oauth2/filter.cc [hunk 4]
```diff
@@ -517,6 +543,36 @@ void OAuth2Filter::finishGetAccessTokenFlow() {
...
+                                  const std::string& data, size_t maxChunkSize = 4000) const {
+  if (data.size() <= maxChunkSize) {
+    // If the data is smaller than the maximum chunk size, no need to split
+    setCookie(headers, key, data);
+    return;
+  }
+
+  if (data.size() > maxChunkSize) {
```
Comment (vikaschoudhary16) [inline source/extensions/filters/http/oauth2/filter.cc:559]: setting chuncked cookies can be moved to a dedicated function `setChunkedCookies` which would return chunk count.

Comment (kanurag94) [reply source/extensions/filters/http/oauth2/filter.cc]: done

File: source/extensions/filters/http/oauth2/filter.cc (hunk 3 above)
Comment (kanurag94) [reply source/extensions/filters/http/oauth2/filter.cc]: done

File: source/extensions/filters/http/oauth2/filter.cc (hunk 1 above)
Comment (kanurag94) [reply source/extensions/filters/http/oauth2/filter.cc]: Most modern browsers have 4096 as the limit, 4000 should suffice all cases. If we choose to take this utility function out to Http::Utility then probably we should have the configuration option, wdyt?

**Comment by kanurag94** [comment]
gentle ping @derekargueta @mattklein123 (maintainers) for review

**Comment by derekargueta** [comment]
https://github.com/envoyproxy/envoy/pull/28867 might inadvertently address this problem as well by producing shorter bearer tokens, so I'm thinking we try to get that merged in first and see if this additional complexity is even needed.

**Comment by kanurag94** [comment]
thanks @derekargueta for mentioning the issue.

IIUC from the PR, I think this may still be an issue with `forward_bearer_token` where we wish to set the token as a part of cookie to be used by the end application. Also, we are using SHA256 so its not reversible to get the token back and set it in a header for upstream. Most of the encoding algorithms will not produce much smaller strings.

**Comment by mattklein123** [comment]
Will wait to hear from @derekargueta to see if we should move forward with this PR or not.

/wait-any

**Comment by spodobetsd** [comment]
The problem is still relevant. Our token contain a lot of groups. I tried to use latest envoy image (dev) with options 
```
disable_id_token_set_cookie: true
disable_access_token_set_cookie: true
disable_refresh_token_set_cookie: true
```
but it didn't help me. 
@derekargueta @mattklein123 could you please get back this PR to review?

**Comment by shashankram** [comment]
> The problem is still relevant. Our token contain a lot of groups. I tried to use latest envoy image (dev) with options
> 
> ```
> disable_id_token_set_cookie: true
> disable_access_token_set_cookie: true
> disable_refresh_token_set_cookie: true
> ```
> 
> but it didn't help me. @derekargueta @mattklein123 could you please get back this PR to review?

@spodobetsd revived cookie chunking in https://github.com/envoyproxy/envoy/pull/43135

**Pull request closed without merging**